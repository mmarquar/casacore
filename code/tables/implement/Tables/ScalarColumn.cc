//# ScalarColumn.cc: Access to a scalar table column with arbitrary data type
//# Copyright (C) 1994,1995,1996,1997
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <aips/Tables/ScalarColumn.h>
#include <aips/Tables/Table.h>
#include <aips/Tables/BaseColumn.h>
#include <aips/Arrays/Vector.h>
#include <aips/Lattices/Slicer.h>
#include <aips/Utilities/ValTypeId.h>
#include <aips/Utilities/String.h>
#include <aips/Tables/TableError.h>


template<class T>
ROScalarColumn<T>::ROScalarColumn()
: ROTableColumn(),
  canAccessColumn_p   (new Bool(False)),
  reaskAccessColumn_p (new Bool(True))
{}

template<class T>
ROScalarColumn<T>::ROScalarColumn (const Table& tab,
				   const String& columnName)
: ROTableColumn (tab, columnName),
  canAccessColumn_p   (new Bool(False)),
  reaskAccessColumn_p (new Bool(True))
{
    checkDataType();
}

template<class T>
ROScalarColumn<T>::ROScalarColumn (const ROTableColumn& column)
: ROTableColumn (column),
  canAccessColumn_p   (new Bool(False)),
  reaskAccessColumn_p (new Bool(True))
{
    checkDataType();
}

template<class T>
ROScalarColumn<T>::ROScalarColumn (const ROScalarColumn<T>& that)
: ROTableColumn (that),
  canAccessColumn_p   (new Bool (*that.canAccessColumn_p)),
  reaskAccessColumn_p (new Bool (*that.reaskAccessColumn_p))
{}

template<class T>
ROTableColumn* ROScalarColumn<T>::clone() const
{
    return new ROScalarColumn<T> (*this);
}

template<class T>
void ROScalarColumn<T>::reference (const ROScalarColumn<T>& that)
{
    ROTableColumn::reference (that);
    *canAccessColumn_p   = *that.canAccessColumn_p;
    *reaskAccessColumn_p = *that.reaskAccessColumn_p;
}

template<class T>
ROScalarColumn<T>::~ROScalarColumn()
{
    delete canAccessColumn_p;
    delete reaskAccessColumn_p;
}

template<class T>
void ROScalarColumn<T>::checkDataType() const
{
    //# Check if the data type matches.
    const ColumnDesc& cd = baseColPtr_p->columnDesc();
    DataType dtype = cd.dataType();
    if (dtype != ValType::getType((T*)0)  ||  !cd.isScalar()) {
	throw (TableInvDT (" in ROScalarColumn ctor for column " + cd.name()));
    }
    if (dtype == TpOther) {
	if (cd.dataTypeId() != valDataTypeId((T*)0)) {
	    throw (TableInvDT (" in ROScalarColumn ctor for column "
			       + cd.name() + "; using data type id "
			       + valDataTypeId((T*)0)
			       + ", expected " + cd.dataTypeId()));
	}
    }
}


template<class T>
Vector<T> ROScalarColumn<T>::getColumn() const
{
    Vector<T> vec;
    getColumn (vec);
    return vec;
}

template<class T>
void ROScalarColumn<T>::getColumn (Vector<T>& vec, Bool resize) const
{
    uInt nrrow = nrow();
    //# Resize the vector if empty; otherwise check its length.
    if (vec.nelements() != nrrow) {
	if (resize  ||  vec.nelements() == 0) {
	    vec.resize (nrrow);
	}else{
	    throw (TableConformanceError("ScalarColumn::getColumn"));
	}
    }
    //# Ask if we can access the column (if that is not known yet).
    if (*reaskAccessColumn_p) {
	*canAccessColumn_p = baseColPtr_p->canAccessScalarColumn
	                                             (*reaskAccessColumn_p);
    }
    //# Access the column if possible.
    //# Otherwise fill the entire vector by looping through all cells.
    if (*canAccessColumn_p) {
	baseColPtr_p->getScalarColumn (&vec);
    }else{
	for (uInt rownr=0; rownr<nrrow; rownr++) {
	    baseColPtr_p->get (rownr, &(vec(rownr)));
	}
    }
}


template<class T>
Vector<T> ROScalarColumn<T>::getColumnRange (const Slicer& rowRange) const
{
    Vector<T> vec;
    getColumnRange (rowRange, vec);
    return vec;
}

template<class T>
void ROScalarColumn<T>::getColumnRange (const Slicer& rowRange,
					Vector<T>& vec, Bool resize) const
{
    uInt nrrow = nrow();
    IPosition shp, blc, trc, inc;
    shp = rowRange.inferShapeFromSource (IPosition(1,nrrow), blc, trc, inc);
    //# When the entire column is accessed, use that function.
    if (shp(0) == Int(nrrow)) {
	getColumn (vec, resize);
	return;
    }
    //# Resize the vector if needed; otherwise check its length.
    nrrow = shp(0);
    if (vec.nelements() != nrrow) {
	if (resize  ||  vec.nelements() == 0) {
	    vec.resize (nrrow);
	}else{
	    throw (TableConformanceError("ScalarColumn::getColumnRange"));
	}
    }
    //# Fill the entire vector by looping through all cells.
    uInt rownr = blc(0);
    uInt incr = inc(0);
    for (uInt i=0; i<nrrow; i++) {
	baseColPtr_p->get (rownr, &(vec(i)));
	rownr += incr;
    }
}



template<class T>
ScalarColumn<T>::ScalarColumn()
: ROTableColumn     (),
  ROScalarColumn<T> (),
  TableColumn       ()
{}

template<class T>
ScalarColumn<T>::ScalarColumn (const Table& tab, const String& columnName)
: ROTableColumn     (tab, columnName),
  ROScalarColumn<T> (tab, columnName),
  TableColumn       (tab, columnName)
{}

template<class T>
ScalarColumn<T>::ScalarColumn (const TableColumn& column)
: ROTableColumn     (column),
  ROScalarColumn<T> (column),
  TableColumn       (column)
{}

template<class T>
ScalarColumn<T>::ScalarColumn (const ScalarColumn<T>& that)
: ROTableColumn     (that),
  ROScalarColumn<T> (that),
  TableColumn       (that)
{}

template<class T>
ROTableColumn* ScalarColumn<T>::clone() const
{
    return new ScalarColumn<T> (*this);
}

template<class T>
void ScalarColumn<T>::reference (const ScalarColumn<T>& that)
    { ROScalarColumn<T>::reference (that); }

template<class T>
ScalarColumn<T>::~ScalarColumn()
{}

template<class T>
void ScalarColumn<T>::put (uInt thisRownr, const ROScalarColumn<T>& that,
			   uInt thatRownr)
{
    put (thisRownr, that(thatRownr));
}

template<class T>
void ScalarColumn<T>::put (uInt thisRownr, const ROTableColumn& that,
			   uInt thatRownr)
{
    T value;
    that.getScalarValue (thatRownr, &value, columnDesc().dataTypeId());
    put (thisRownr, value);
}

template<class T>
void ScalarColumn<T>::putColumn (const Vector<T>& vec)
{
    uInt nrrow = nrow();
    //# Check the vector length.
    if (vec.nelements() != nrrow) {
	throw (TableConformanceError("ScalarColumn::putColumn(Vector&)"));
    }
    //# Ask if we can access the column (if that is not known yet).
    if (*reaskAccessColumn_p) {
	*canAccessColumn_p = baseColPtr_p->canAccessScalarColumn
	                                             (*reaskAccessColumn_p);
    }
    //# Access the column if possible.
    //# Otherwise put the entire vector by looping through all cells.
    if (*canAccessColumn_p) {
	baseColPtr_p->putScalarColumn (&vec);
    }else{
	for (uInt rownr=0; rownr<nrrow; rownr++) {
	    baseColPtr_p->put (rownr, &(vec(rownr)));
	}
    }
}

template<class T>
void ScalarColumn<T>::putColumnRange (const Slicer& rowRange,
				      const Vector<T>& vec)
{
    uInt nrrow = nrow();
    IPosition shp, blc, trc, inc;
    shp = rowRange.inferShapeFromSource (IPosition(1,nrrow), blc, trc, inc);
    //# When the entire column is accessed, use that function.
    if (shp(0) == Int(nrrow)) {
	putColumn (vec);
	return;
    }
    //# Check the vector length.
    nrrow = shp(0);
    if (vec.nelements() != nrrow) {
	throw (TableConformanceError("ScalarColumn::putColumnRange"));
    }
    //# Put the entire vector by looping through all cells.
    uInt rownr = blc(0);
    uInt incr = inc(0);
    for (uInt i=0; i<nrrow; i++) {
	baseColPtr_p->put (rownr, &(vec(i)));
	rownr += incr;
    }
}


//#// This is a very simple implementation.
//#// Ultimately this must be done more direct via the data manager.
template<class T>
void ScalarColumn<T>::fillColumn (const T& value)
{
    uInt nrrow = nrow();
    for (uInt i=0; i<nrrow; i++) {
	put (i, value);
    }
}

template<class T>
void ScalarColumn<T>::putColumn (const ROScalarColumn<T>& that)
{
    //# Check the column lengths.
    uInt nrrow = nrow();
    if (nrrow != that.nrow()) {
	throw (TableConformanceError ("ScalarColumn<T>::putColumn"));
    }
    for (uInt i=0; i<nrrow; i++) {
	put (i, that, i);
    }
}
