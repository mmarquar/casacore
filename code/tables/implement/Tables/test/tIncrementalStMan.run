#!/bin/sh
#-----------------------------------------------------------------------------
# Script to invoke tIncrementalStMan and verify its output.
#=============================================================================

exec assay tIncrementalStMan "1000 && \
     tIncrementalStMan 2000 && \
     tIncrementalStMan 3000 && \
     tIncrementalStMan 4000 && \
     tIncrementalStMan 5000 && \
     tIncrementalStMan 0"
