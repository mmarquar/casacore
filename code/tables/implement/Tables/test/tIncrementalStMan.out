10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 0, 0, 0, 1, 1, 1, 2, 2, 2, 3
10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 0, 0, 0, 1, 1, 1, 2, 2, 2, 3, 
, , , , , , , , , , abca, abca, abca, abcaa, abcaa, abcaa, abcaaa, abcaaa, abcaaa, abcaaaa
11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 1, 1, 1, 2, 2, 2, 3, 3, 3, 4, 
arr4 = Axis Lengths: [1, 20]  (NB: Matrix in Row/Column order)
[z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z]

arr4 = Axis Lengths: [1, 20]  (NB: Matrix in Row/Column order)
[z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z]

arr4 = Axis Lengths: [1, 20]  (NB: Matrix in Row/Column order)
[z, a, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z]

arr4 = Axis Lengths: [1, 20]  (NB: Matrix in Row/Column order)
[z, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a]

#Rows 20
(-1.4013e-45, -1.4013e-45), (2, 0), (2, 0), (4, 0), (4, 0), (6, 0), (6, 0), (8, 0), (8, 0), (10, 0), (10, 0), (12, 0), (12, 0), (14, 0), (14, 0), (16, 0), (16, 0), (18, 0), (18, 0), (20, 0)
0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19
11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 1, 1, 1, 2, 2, 2, 3, 3, 3, 4
, , , , , , , , , , abca, abca, abca, abcaa, abcaa, abcaa, abcaaa, abcaaa, abcaaa, abcaaaa
Ndim=3 Axis Lengths: [2, 3, 4] 
[0, 0, 0][432, 433]
[0, 1, 0][434, 435]
[0, 2, 0][436, 437]
[0, 0, 1][438, 439]
[0, 1, 1][440, 441]
[0, 2, 1][442, 443]
[0, 0, 2][444, 445]
[0, 1, 2][446, 447]
[0, 2, 2][448, 449]
[0, 0, 3][450, 451]
[0, 1, 3][452, 453]
[0, 2, 3][454, 455]

Ndim=3 Axis Lengths: [2, 3, 4] 
[0, 0, 0][408, 409]
[0, 1, 0][410, 411]
[0, 2, 0][412, 413]
[0, 0, 1][414, 415]
[0, 1, 1][416, 417]
[0, 2, 1][418, 419]
[0, 0, 2][420, 421]
[0, 1, 2][422, 423]
[0, 2, 2][424, 425]
[0, 0, 3][426, 427]
[0, 1, 3][428, 429]
[0, 2, 3][430, 431]

Ndim=3 Axis Lengths: [2, 3, 4] 
[0, 0, 0][408, 409]
[0, 1, 0][410, 411]
[0, 2, 0][412, 413]
[0, 0, 1][414, 415]
[0, 1, 1][416, 417]
[0, 2, 1][418, 419]
[0, 0, 2][420, 421]
[0, 1, 2][422, 423]
[0, 2, 2][424, 425]
[0, 0, 3][426, 427]
[0, 1, 3][428, 429]
[0, 2, 3][430, 431]

10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 0, 0, 0, 1, 1, 1, 2, 2, 2, 3
10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 0, 0, 0, 1, 1, 1, 2, 2, 2, 3, 
, , , , , , , , , , abca, abca, abca, abcaa, abcaa, abcaa, abcaaa, abcaaa, abcaaa, abcaaaa
11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 1, 1, 1, 2, 2, 2, 3, 3, 3, 4, 
arr4 = Axis Lengths: [1, 20]  (NB: Matrix in Row/Column order)
[z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z]

arr4 = Axis Lengths: [1, 20]  (NB: Matrix in Row/Column order)
[z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z]

arr4 = Axis Lengths: [1, 20]  (NB: Matrix in Row/Column order)
[z, a, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z]

arr4 = Axis Lengths: [1, 20]  (NB: Matrix in Row/Column order)
[z, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a]

#Rows 20
(-1.4013e-45, -1.4013e-45), (2, 0), (2, 0), (4, 0), (4, 0), (6, 0), (6, 0), (8, 0), (8, 0), (10, 0), (10, 0), (12, 0), (12, 0), (14, 0), (14, 0), (16, 0), (16, 0), (18, 0), (18, 0), (20, 0)
0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19
11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 1, 1, 1, 2, 2, 2, 3, 3, 3, 4
, , , , , , , , , , abca, abca, abca, abcaa, abcaa, abcaa, abcaaa, abcaaa, abcaaa, abcaaaa
Ndim=3 Axis Lengths: [2, 3, 4] 
[0, 0, 0][432, 433]
[0, 1, 0][434, 435]
[0, 2, 0][436, 437]
[0, 0, 1][438, 439]
[0, 1, 1][440, 441]
[0, 2, 1][442, 443]
[0, 0, 2][444, 445]
[0, 1, 2][446, 447]
[0, 2, 2][448, 449]
[0, 0, 3][450, 451]
[0, 1, 3][452, 453]
[0, 2, 3][454, 455]

Ndim=3 Axis Lengths: [2, 3, 4] 
[0, 0, 0][408, 409]
[0, 1, 0][410, 411]
[0, 2, 0][412, 413]
[0, 0, 1][414, 415]
[0, 1, 1][416, 417]
[0, 2, 1][418, 419]
[0, 0, 2][420, 421]
[0, 1, 2][422, 423]
[0, 2, 2][424, 425]
[0, 0, 3][426, 427]
[0, 1, 3][428, 429]
[0, 2, 3][430, 431]

Ndim=3 Axis Lengths: [2, 3, 4] 
[0, 0, 0][408, 409]
[0, 1, 0][410, 411]
[0, 2, 0][412, 413]
[0, 0, 1][414, 415]
[0, 1, 1][416, 417]
[0, 2, 1][418, 419]
[0, 0, 2][420, 421]
[0, 1, 2][422, 423]
[0, 2, 2][424, 425]
[0, 0, 3][426, 427]
[0, 1, 3][428, 429]
[0, 2, 3][430, 431]

10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 0, 0, 0, 1, 1, 1, 2, 2, 2, 3
10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 0, 0, 0, 1, 1, 1, 2, 2, 2, 3, 
, , , , , , , , , , abca, abca, abca, abcaa, abcaa, abcaa, abcaaa, abcaaa, abcaaa, abcaaaa
11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 1, 1, 1, 2, 2, 2, 3, 3, 3, 4, 
arr4 = Axis Lengths: [1, 20]  (NB: Matrix in Row/Column order)
[z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z]

arr4 = Axis Lengths: [1, 20]  (NB: Matrix in Row/Column order)
[z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z]

arr4 = Axis Lengths: [1, 20]  (NB: Matrix in Row/Column order)
[z, a, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z]

arr4 = Axis Lengths: [1, 20]  (NB: Matrix in Row/Column order)
[z, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a]

#Rows 20
(-1.4013e-45, -1.4013e-45), (2, 0), (2, 0), (4, 0), (4, 0), (6, 0), (6, 0), (8, 0), (8, 0), (10, 0), (10, 0), (12, 0), (12, 0), (14, 0), (14, 0), (16, 0), (16, 0), (18, 0), (18, 0), (20, 0)
0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19
11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 1, 1, 1, 2, 2, 2, 3, 3, 3, 4
, , , , , , , , , , abca, abca, abca, abcaa, abcaa, abcaa, abcaaa, abcaaa, abcaaa, abcaaaa
Ndim=3 Axis Lengths: [2, 3, 4] 
[0, 0, 0][432, 433]
[0, 1, 0][434, 435]
[0, 2, 0][436, 437]
[0, 0, 1][438, 439]
[0, 1, 1][440, 441]
[0, 2, 1][442, 443]
[0, 0, 2][444, 445]
[0, 1, 2][446, 447]
[0, 2, 2][448, 449]
[0, 0, 3][450, 451]
[0, 1, 3][452, 453]
[0, 2, 3][454, 455]

Ndim=3 Axis Lengths: [2, 3, 4] 
[0, 0, 0][408, 409]
[0, 1, 0][410, 411]
[0, 2, 0][412, 413]
[0, 0, 1][414, 415]
[0, 1, 1][416, 417]
[0, 2, 1][418, 419]
[0, 0, 2][420, 421]
[0, 1, 2][422, 423]
[0, 2, 2][424, 425]
[0, 0, 3][426, 427]
[0, 1, 3][428, 429]
[0, 2, 3][430, 431]

Ndim=3 Axis Lengths: [2, 3, 4] 
[0, 0, 0][408, 409]
[0, 1, 0][410, 411]
[0, 2, 0][412, 413]
[0, 0, 1][414, 415]
[0, 1, 1][416, 417]
[0, 2, 1][418, 419]
[0, 0, 2][420, 421]
[0, 1, 2][422, 423]
[0, 2, 2][424, 425]
[0, 0, 3][426, 427]
[0, 1, 3][428, 429]
[0, 2, 3][430, 431]

10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 0, 0, 0, 1, 1, 1, 2, 2, 2, 3
10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 0, 0, 0, 1, 1, 1, 2, 2, 2, 3, 
, , , , , , , , , , abca, abca, abca, abcaa, abcaa, abcaa, abcaaa, abcaaa, abcaaa, abcaaaa
11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 1, 1, 1, 2, 2, 2, 3, 3, 3, 4, 
arr4 = Axis Lengths: [1, 20]  (NB: Matrix in Row/Column order)
[z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z]

arr4 = Axis Lengths: [1, 20]  (NB: Matrix in Row/Column order)
[z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z]

arr4 = Axis Lengths: [1, 20]  (NB: Matrix in Row/Column order)
[z, a, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z]

arr4 = Axis Lengths: [1, 20]  (NB: Matrix in Row/Column order)
[z, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a]

#Rows 20
(-1.4013e-45, -1.4013e-45), (2, 0), (2, 0), (4, 0), (4, 0), (6, 0), (6, 0), (8, 0), (8, 0), (10, 0), (10, 0), (12, 0), (12, 0), (14, 0), (14, 0), (16, 0), (16, 0), (18, 0), (18, 0), (20, 0)
0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19
11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 1, 1, 1, 2, 2, 2, 3, 3, 3, 4
, , , , , , , , , , abca, abca, abca, abcaa, abcaa, abcaa, abcaaa, abcaaa, abcaaa, abcaaaa
Ndim=3 Axis Lengths: [2, 3, 4] 
[0, 0, 0][432, 433]
[0, 1, 0][434, 435]
[0, 2, 0][436, 437]
[0, 0, 1][438, 439]
[0, 1, 1][440, 441]
[0, 2, 1][442, 443]
[0, 0, 2][444, 445]
[0, 1, 2][446, 447]
[0, 2, 2][448, 449]
[0, 0, 3][450, 451]
[0, 1, 3][452, 453]
[0, 2, 3][454, 455]

Ndim=3 Axis Lengths: [2, 3, 4] 
[0, 0, 0][408, 409]
[0, 1, 0][410, 411]
[0, 2, 0][412, 413]
[0, 0, 1][414, 415]
[0, 1, 1][416, 417]
[0, 2, 1][418, 419]
[0, 0, 2][420, 421]
[0, 1, 2][422, 423]
[0, 2, 2][424, 425]
[0, 0, 3][426, 427]
[0, 1, 3][428, 429]
[0, 2, 3][430, 431]

Ndim=3 Axis Lengths: [2, 3, 4] 
[0, 0, 0][408, 409]
[0, 1, 0][410, 411]
[0, 2, 0][412, 413]
[0, 0, 1][414, 415]
[0, 1, 1][416, 417]
[0, 2, 1][418, 419]
[0, 0, 2][420, 421]
[0, 1, 2][422, 423]
[0, 2, 2][424, 425]
[0, 0, 3][426, 427]
[0, 1, 3][428, 429]
[0, 2, 3][430, 431]

10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 0, 0, 0, 1, 1, 1, 2, 2, 2, 3
10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 0, 0, 0, 1, 1, 1, 2, 2, 2, 3, 
, , , , , , , , , , abca, abca, abca, abcaa, abcaa, abcaa, abcaaa, abcaaa, abcaaa, abcaaaa
11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 1, 1, 1, 2, 2, 2, 3, 3, 3, 4, 
arr4 = Axis Lengths: [1, 20]  (NB: Matrix in Row/Column order)
[z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z]

arr4 = Axis Lengths: [1, 20]  (NB: Matrix in Row/Column order)
[z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z]

arr4 = Axis Lengths: [1, 20]  (NB: Matrix in Row/Column order)
[z, a, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z]

arr4 = Axis Lengths: [1, 20]  (NB: Matrix in Row/Column order)
[z, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a]

#Rows 20
(-1.4013e-45, -1.4013e-45), (2, 0), (2, 0), (4, 0), (4, 0), (6, 0), (6, 0), (8, 0), (8, 0), (10, 0), (10, 0), (12, 0), (12, 0), (14, 0), (14, 0), (16, 0), (16, 0), (18, 0), (18, 0), (20, 0)
0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19
11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 1, 1, 1, 2, 2, 2, 3, 3, 3, 4
, , , , , , , , , , abca, abca, abca, abcaa, abcaa, abcaa, abcaaa, abcaaa, abcaaa, abcaaaa
Ndim=3 Axis Lengths: [2, 3, 4] 
[0, 0, 0][432, 433]
[0, 1, 0][434, 435]
[0, 2, 0][436, 437]
[0, 0, 1][438, 439]
[0, 1, 1][440, 441]
[0, 2, 1][442, 443]
[0, 0, 2][444, 445]
[0, 1, 2][446, 447]
[0, 2, 2][448, 449]
[0, 0, 3][450, 451]
[0, 1, 3][452, 453]
[0, 2, 3][454, 455]

Ndim=3 Axis Lengths: [2, 3, 4] 
[0, 0, 0][408, 409]
[0, 1, 0][410, 411]
[0, 2, 0][412, 413]
[0, 0, 1][414, 415]
[0, 1, 1][416, 417]
[0, 2, 1][418, 419]
[0, 0, 2][420, 421]
[0, 1, 2][422, 423]
[0, 2, 2][424, 425]
[0, 0, 3][426, 427]
[0, 1, 3][428, 429]
[0, 2, 3][430, 431]

Ndim=3 Axis Lengths: [2, 3, 4] 
[0, 0, 0][408, 409]
[0, 1, 0][410, 411]
[0, 2, 0][412, 413]
[0, 0, 1][414, 415]
[0, 1, 1][416, 417]
[0, 2, 1][418, 419]
[0, 0, 2][420, 421]
[0, 1, 2][422, 423]
[0, 2, 2][424, 425]
[0, 0, 3][426, 427]
[0, 1, 3][428, 429]
[0, 2, 3][430, 431]

10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 0, 0, 0, 1, 1, 1, 2, 2, 2, 3
10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 0, 0, 0, 1, 1, 1, 2, 2, 2, 3, 
, , , , , , , , , , abca, abca, abca, abcaa, abcaa, abcaa, abcaaa, abcaaa, abcaaa, abcaaaa
11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 1, 1, 1, 2, 2, 2, 3, 3, 3, 4, 
arr4 = Axis Lengths: [1, 20]  (NB: Matrix in Row/Column order)
[z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z]

arr4 = Axis Lengths: [1, 20]  (NB: Matrix in Row/Column order)
[z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z]

arr4 = Axis Lengths: [1, 20]  (NB: Matrix in Row/Column order)
[z, a, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z]

arr4 = Axis Lengths: [1, 20]  (NB: Matrix in Row/Column order)
[z, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a]

#Rows 20
(-1.4013e-45, -1.4013e-45), (2, 0), (2, 0), (4, 0), (4, 0), (6, 0), (6, 0), (8, 0), (8, 0), (10, 0), (10, 0), (12, 0), (12, 0), (14, 0), (14, 0), (16, 0), (16, 0), (18, 0), (18, 0), (20, 0)
0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19
11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 1, 1, 1, 2, 2, 2, 3, 3, 3, 4
, , , , , , , , , , abca, abca, abca, abcaa, abcaa, abcaa, abcaaa, abcaaa, abcaaa, abcaaaa
Ndim=3 Axis Lengths: [2, 3, 4] 
[0, 0, 0][432, 433]
[0, 1, 0][434, 435]
[0, 2, 0][436, 437]
[0, 0, 1][438, 439]
[0, 1, 1][440, 441]
[0, 2, 1][442, 443]
[0, 0, 2][444, 445]
[0, 1, 2][446, 447]
[0, 2, 2][448, 449]
[0, 0, 3][450, 451]
[0, 1, 3][452, 453]
[0, 2, 3][454, 455]

Ndim=3 Axis Lengths: [2, 3, 4] 
[0, 0, 0][408, 409]
[0, 1, 0][410, 411]
[0, 2, 0][412, 413]
[0, 0, 1][414, 415]
[0, 1, 1][416, 417]
[0, 2, 1][418, 419]
[0, 0, 2][420, 421]
[0, 1, 2][422, 423]
[0, 2, 2][424, 425]
[0, 0, 3][426, 427]
[0, 1, 3][428, 429]
[0, 2, 3][430, 431]

Ndim=3 Axis Lengths: [2, 3, 4] 
[0, 0, 0][408, 409]
[0, 1, 0][410, 411]
[0, 2, 0][412, 413]
[0, 0, 1][414, 415]
[0, 1, 1][416, 417]
[0, 2, 1][418, 419]
[0, 0, 2][420, 421]
[0, 1, 2][422, 423]
[0, 2, 2][424, 425]
[0, 0, 3][426, 427]
[0, 1, 3][428, 429]
[0, 2, 3][430, 431]

